File: MeatGrinder/Repositories/TaskRepository.cs
﻿using System.Collections.Generic;
using System.Linq;
using MeatGrinder.Models;
using MeatGrinder.Services;

namespace MeatGrinder.Repositories
{
    public class TaskRepository
    {
        private readonly MeatGrinderEntities _db = new MeatGrinderEntities();

        public Task GetByID(int id)
        {
            return _db.Tasks.FirstOrDefault(m => m.ID == id);
        }
        public List<Task> GetAllByGoalID(int goalId)
        {
            int userID = CookieService.GetUserID();
            return _db.Tasks.Where(m => m.GoalID == goalId &&
                                        m.ParentTaskID == null &&
                                        m.UserID == userID).ToList();
        }
        public List<Task> GetAllByTaskID(int taskId)
        {
            int userID = CookieService.GetUserID();
            return _db.Tasks.Where(m => m.ParentTaskID == taskId &&
                                        m.UserID == userID).ToList();
        }
        public int GetChildTaskCount(int taskID)
        {
            return _db.Tasks.Count(m => m.ParentTaskID == taskID);
        }
        public void Create(Task task)
        {
            task.UserID = CookieService.GetUserID();
            task.IsComplete = false;
            _db.Tasks.Add(task);
            _db.SaveChanges();

            if (task.ParentTaskID != null)
            {
                SetTaskNotComplete(task.ParentTaskID.Value);
            }
            else
            {
                var goalRepository = new GoalRepository();

                goalRepository.SetGoalNotComplete(task.GoalID);
            }
        }
        public void SetTaskNotComplete(int taskId)
        {
            var task = _db.Tasks.FirstOrDefault(m => m.ID == taskId);

            if (task != null && task.IsComplete)
            {
                task.IsComplete = false;
                _db.SaveChanges();
            }
        }
        public void Delete(Task task)
        {
            task = _db.Tasks.FirstOrDefault(m => m.ID == task.ID);

            if (task != null) 
                DeleteTask(task);

            _db.SaveChanges();
        }

        private void DeleteTask(Task task)
        {
            var childTasks = _db.Tasks.Where(m => m.ParentTaskID == task.ID);

            foreach (var childTask in childTasks)
            {
                DeleteTask(childTask);
            }

            _db.Tasks.Remove(task);
        }
    }
}

File: MeatGrinder.Web/Repositories/TaskRepository.fs
namespace MeatGrinder.Web.Repositories
open System.Linq
open Microsoft.FSharp.Linq.NullableOperators

open MeatGrinder.Orm

type TaskRepository(db:MeatGrinderEntities, getUserId:unit->int option) =
    member x.GetById id = db.Tasks.FirstOrDefault(fun m-> m.Id=id)
    member x.GetAllByGoalId goalId = 
        let id = getUserId().Value
        db.Tasks.Where(fun m->m.GoalId=goalId && m.ParentTaskID.HasValue=false && m.UserId= id) |> Seq.toList
    member x.GetAllByTaskId (taskId:int) =
        let userId = getUserId().Value
        db.Tasks.Where(fun m->m.ParentTaskID ?= taskId && m.UserId=userId) |> Seq.toList
    member x.GetChildTaskCount taskId =
        db.Tasks.Count (fun m-> m.ParentTaskID ?= taskId)
    member x.SetTaskNotComplete taskId =
        let task = db.Tasks.FirstOrDefault(fun m->m.Id=taskId)
        if task.IsComplete then
            task.IsComplete <- false
            db.SaveChanges() |> ignore
    member x.Create (task:Task) setGoalNotComplete =
        task.UserId <- getUserId().Value
        task.IsComplete <- false
        db.Tasks.Add task |> ignore
        db.SaveChanges() |> ignore
        if task.ParentTaskID.HasValue
        then x.SetTaskNotComplete task.ParentTaskID.Value
        else setGoalNotComplete task.GoalId
    
    member x.DeleteTask (task:Task) =
        let rec delete (parentTask:Task) =
            db.Tasks.Where(fun m->m.ParentTaskID ?= parentTask.Id) 
                |> Seq.toList
                |> Seq.iter (fun t-> x.DeleteTask t)
        delete task
        db.Tasks.Remove(task) |> ignore
        
    member x.Delete taskId =
        let task = db.Tasks.FirstOrDefault(fun m->m.Id=taskId)

        if task.Id>0 then
            x.DeleteTask task
            db.SaveChanges() |> ignore
    member x.UpdateChildTaskCounts (tasks:Task list) =
        let upd (t:Task) = 
            let childTaskCount = x.GetChildTaskCount(t.Id)
            t.ChildTaskCount <- Some(childTaskCount)
            if t.ChildTaskCount.Value>0 then t.Description <- sprintf "%s(%i)" t.Description childTaskCount
        tasks |> Seq.iter upd
